### PR TITLE
[core]Migrate MatMul operator to new API

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Migrate `MatMul` operator to new API.
 - Minor refactor operator to reduce binary size.

### Tickets:
 - [CVS-118624](https://jira.devtools.intel.com/browse/CVS-118624)

### CODE REVIEW

File: src/core/src/op/matmul.cpp [hunk 1]
```diff
@@ -117,8 +106,10 @@ void ngraph::op::v0::MatMul::validate_and_infer_types() {
                           get_input_element_type(1),
                           ").");
 
-    const auto &A_shape = get_input_partial_shape(0), B_shape = get_input_partial_shape(1);
-    std::vector<ov::PartialShape> input_shapes = {A_shape, B_shape};
-    std::vector<ov::PartialShape> output_shapes = shape_infer(this, input_shapes);
+    const auto &A_shape = get_input_partial_shape(0), &B_shape = get_input_partial_shape(1);
```
Comment (t-jankowski) [inline src/core/src/op/matmul.cpp:109]: Should it be split into two lines?

**Review by t-jankowski — APPROVED** [review]

**Review by vurusovs — APPROVED** [review]

File: src/core/src/op/matmul.cpp [hunk 2]
```diff
@@ -2,110 +2,99 @@
...
+    return std::make_shared<MatMul>(new_args.at(0), new_args.at(1), m_transpose_a, m_transpose_b);
 }
-}  // namespace
-}  // namespace matmul
 
-bool op::MatMul::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const {
+bool MatMul::evaluate(TensorVector& outputs, const TensorVector& inputs) const {
     OV_OP_SCOPE(v0_MatMul_evaluate);
-    return matmul::evaluate_matmul(this, inputs[0], inputs[1], outputs[0]);
+    OPENVINO_ASSERT(outputs.size() == 1);
```
Comment (vurusovs) [inline src/core/src/op/matmul.cpp:65]: do we need to check that there is 2 inputs?

Comment (praasz) [reply src/core/src/op/matmul.cpp:65]: Is done in `shape_infer`` function.

File: src/core/src/op/matmul.cpp (hunk 1 above)
Comment (praasz) [reply src/core/src/op/matmul.cpp:109]: Changed, as in current version is easy make mistake between value and reference.

**Pull request merged**